File: dev-kit/dev_kit/agent/crypto.py
"""
dev-kit/dev_kit/agent/crypto.py

Server-side RSA-4096 key pair management and hybrid decryption for secrets
sent encrypted from the browser via the Web Crypto API.

Key lifecycle:
  - On first run, generate a 4096-bit RSA key pair and persist the private
    key to ~/.devkit/server_key.pem (mode 0o600, directory 0o700).
  - On subsequent runs, load the persisted key.
  - The public key is exposed via GET /api/deploy/public-key so the browser
    can encrypt secrets before sending them over HTTP.

Encryption scheme (mirrors browser Web Crypto):
  Browser side:
    1. Generate a random 32-byte AES-256-GCM key.
    2. Encrypt the plaintext secret with AES-256-GCM (12-byte random IV).
    3. Encrypt the AES key with RSA-OAEP (SHA-256).
    4. Send { encrypted_key, iv, encrypted_value } — all base64-encoded.

  Server side (this module):
    1. Decrypt the AES key with the RSA private key (OAEP + SHA-256).
    2. Decrypt the value with AES-256-GCM using the recovered key and IV.
"""

from __future__ import annotations

import base64
import logging
import os
from pathlib import Path

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding, rsa
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

logger = logging.getLogger(__name__)

_KEY_PATH = Path.home() / ".devkit" / "server_key.pem"


def _load_or_generate_key() -> rsa.RSAPrivateKey:
    """Load the server private key from disk, generating it if absent.

    The key file is created at ~/.devkit/server_key.pem with mode 0o600.
    The directory is created with mode 0o700 if it does not exist.

    Returns:
        The RSA-4096 private key instance.
    """
    if _KEY_PATH.exists():
        with _KEY_PATH.open("rb") as f:
            key = serialization.load_pem_private_key(f.read(), password=None)
        logger.info(
            "crypto.key_loaded",
            extra={
                "operation": "crypto.load_key",
                "status": "success",
                "path": str(_KEY_PATH),
                "latency_ms": 0,
            },
        )
        return key  # type: ignore[return-value]

    key = rsa.generate_private_key(public_exponent=65537, key_size=4096)
    _KEY_PATH.parent.mkdir(mode=0o700, parents=True, exist_ok=True)
    with _KEY_PATH.open("wb") as f:
        f.write(
            key.private_bytes(
                encoding=serialization.Encoding.PEM,
                format=serialization.PrivateFormat.TraditionalOpenSSL,
                encryption_algorithm=serialization.NoEncryption(),
            )
        )
    _KEY_PATH.chmod(0o600)
    logger.info(
        "crypto.key_generated",
        extra={
            "operation": "crypto.generate_key",
            "status": "success",
            "path": str(_KEY_PATH),
            "latency_ms": 0,
        },
    )
    return key


_PRIVATE_KEY: rsa.RSAPrivateKey = _load_or_generate_key()


def get_public_key_spki_b64() -> str:
    """Return the server public key in base64-encoded DER/SPKI format.

    The browser uses this with ``SubtleCrypto.importKey('spki', ...)`` to
    import the key for RSA-OAEP encryption.

    Returns:
        Base64 string of the DER-encoded SubjectPublicKeyInfo structure.
    """
    spki = _PRIVATE_KEY.public_key().public_bytes(
        encoding=serialization.Encoding.DER,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    )
    return base64.b64encode(spki).decode()


def decrypt_secret(payload: dict) -> str:
    """Decrypt a single hybrid-encrypted secret from the browser.

    The payload must contain the three fields produced by the browser's
    ``encryptSecret()`` function in ``src/crypto.js``.

    Args:
        payload: Dict with keys:
            ``encrypted_key``   — base64 RSA-OAEP ciphertext of the AES key
            ``iv``              — base64 12-byte AES-GCM nonce
            ``encrypted_value`` — base64 AES-256-GCM ciphertext of the secret

    Returns:
        Plaintext secret string.

    Raises:
        KeyError: If any required key is missing from payload.
        ValueError: If decryption fails (wrong key or corrupted payload).
    """
    encrypted_key = base64.b64decode(payload["encrypted_key"])
    iv = base64.b64decode(payload["iv"])
    encrypted_value = base64.b64decode(payload["encrypted_value"])

    aes_key = _PRIVATE_KEY.decrypt(
        encrypted_key,
        padding.OAEP(
            mgf=padding.MGF1(algorithm=hashes.SHA256()),
            algorithm=hashes.SHA256(),
            label=None,
        ),
    )

    aesgcm = AESGCM(aes_key)
    return aesgcm.decrypt(iv, encrypted_value, None).decode()


def decrypt_secrets_dict(encrypted: dict) -> dict:
    """Recursively decrypt a dict of hybrid-encrypted secrets.

    String values that are dicts with ``encrypted_key`` are decrypted.
    Nested dicts (e.g. ``tool_secrets``) are recursed into.
    Empty strings are passed through unchanged.

    Args:
        encrypted: Nested dict where leaf values are either empty strings
            or cipher-payload dicts ``{encrypted_key, iv, encrypted_value}``.

    Returns:
        Dict with the same structure but all cipher payloads replaced by
        their plaintext strings.
    """
    result: dict = {}
    for key, value in encrypted.items():
        if isinstance(value, dict) and "encrypted_key" in value:
            result[key] = decrypt_secret(value)
        elif isinstance(value, dict):
            result[key] = decrypt_secrets_dict(value)
        else:
            result[key] = value  # empty string or non-secret passthrough
    return result

File: dev-kit/frontend/src/crypto.js
/**
 * dev-kit/frontend/src/crypto.js
 *
 * Browser-side hybrid encryption using the Web Crypto API.
 *
 * Encryption scheme:
 *   1. Fetch the server's RSA-4096 public key (SPKI/DER base64) once.
 *   2. For each secret value:
 *      a. Generate a fresh random AES-256-GCM key.
 *      b. Encrypt the value with AES-256-GCM (12-byte random IV).
 *      c. Encrypt the AES key with RSA-OAEP (SHA-256).
 *      d. Return { encrypted_key, iv, encrypted_value } — all base64.
 *
 * The server decrypts with crypto.py:decrypt_secret().
 */

let _cachedPublicKey = null

/**
 * Whether the Web Crypto SubtleCrypto API is available.
 *
 * `window.crypto.subtle` is only exposed in secure contexts (HTTPS or
 * localhost). When the dev-kit is served over plain HTTP from a remote host
 * (e.g. a VM IP), it is `undefined` and any call to importKey/encrypt throws.
 *
 * @returns {boolean}
 */
export function isSubtleCryptoAvailable() {
  return typeof window !== 'undefined'
    && !!window.crypto
    && !!window.crypto.subtle
    && typeof window.crypto.subtle.importKey === 'function'
}

/**
 * Build the secrets payload for deploy endpoints.
 *
 * In a secure context, encrypts every secret with hybrid RSA-OAEP + AES-GCM
 * and returns `{ encrypted_secrets }`. In a non-secure context where
 * SubtleCrypto is unavailable, falls back to `{ secrets }` (plaintext); the
 * dev-kit backend accepts both shapes.
 *
 * @param {Object} secretsDict - Flat or nested object of secret strings.
 * @returns {Promise<Object>} Either `{ encrypted_secrets }` or `{ secrets }`.
 */
export async function buildSecretsPayload(secretsDict) {
  const dict = secretsDict || {}
  if (isSubtleCryptoAvailable()) {
    return { encrypted_secrets: await encryptSecretsDict(dict) }
  }
  return { secrets: dict }
}

/**
 * Fetch and import the server's RSA public key.
 * Result is cached for the lifetime of the page.
 *
 * @returns {Promise<CryptoKey>} The imported RSA-OAEP public key.
 */
export async function fetchPublicKey() {
  if (_cachedPublicKey) return _cachedPublicKey
  const res = await fetch('/api/deploy/public-key')
  const { public_key } = await res.json()
  const keyBytes = Uint8Array.from(atob(public_key), c => c.charCodeAt(0))
  _cachedPublicKey = await window.crypto.subtle.importKey(
    'spki',
    keyBytes,
    { name: 'RSA-OAEP', hash: 'SHA-256' },
    false,
    ['encrypt']
  )
  return _cachedPublicKey
}

/**
 * Encrypt a single plaintext string with hybrid RSA-OAEP + AES-256-GCM.
 *
 * @param {string} value - The plaintext secret to encrypt.
 * @returns {Promise<{encrypted_key: string, iv: string, encrypted_value: string}>}
 *   All values are base64-encoded strings suitable for JSON serialisation.
 */
export async function encryptSecret(value) {
  const publicKey = await fetchPublicKey()

  // Generate a fresh 256-bit AES key for this secret
  const aesKey = await window.crypto.subtle.generateKey(
    { name: 'AES-GCM', length: 256 },
    true,
    ['encrypt']
  )
  const aesKeyRaw = await window.crypto.subtle.exportKey('raw', aesKey)

  // Encrypt the secret value with AES-256-GCM
  const iv = window.crypto.getRandomValues(new Uint8Array(12))
  const encryptedValue = await window.crypto.subtle.encrypt(
    { name: 'AES-GCM', iv },
    aesKey,
    new TextEncoder().encode(value)
  )

  // Encrypt the AES key with RSA-OAEP (the only thing RSA touches)
  const encryptedKey = await window.crypto.subtle.encrypt(
    { name: 'RSA-OAEP' },
    publicKey,
    aesKeyRaw
  )

  const toBase64 = (buf) =>
    btoa(String.fromCharCode(...new Uint8Array(buf)))

  return {
    encrypted_key: toBase64(encryptedKey),
    iv: toBase64(iv),
    encrypted_value: toBase64(encryptedValue),
  }
}

/**
 * Recursively encrypt every non-empty string value in a secrets dict.
 *
 * Nested objects (e.g. tool_secrets) are recursed into.
 * Empty strings are passed through unchanged (no field to encrypt).
 *
 * @param {Object} secretsDict - Flat or nested object of secret strings.
 * @returns {Promise<Object>} Same structure with string values replaced by
 *   cipher-payload objects { encrypted_key, iv, encrypted_value }.
 */
export async function encryptSecretsDict(secretsDict) {
  const result = {}
  for (const [key, value] of Object.entries(secretsDict)) {
    if (typeof value === 'string' && value !== '') {
      result[key] = await encryptSecret(value)
    } else if (value !== null && typeof value === 'object') {
      result[key] = await encryptSecretsDict(value)
    } else {
      result[key] = value  // empty string or null — pass through
    }
  }
  return result
}
